Signature validation Failure example

conradoGen opened the pull request:
In here we can do 2 things:
* get the console.log output of the signed and verified  in nodejs and try to validate in the playground: fail
* conform that the base64 constants (pub, msg, sig) copied from the playground does not verify in nodejs while does in the playground

conradoGen committed:
cross validation example

conradoGen committed:
new pk, sk, message

conradoGen committed:
updated example

conradoGen committed:
reorganized

conradoGen committed:
testing conversions

conradoGen committed:
MVP for error validations

jboillot commented on index.js:
```diff
@@ -4,12 +4,52 @@ import assert from 'node:assert/strict'
+
+    const pGpubUint8Array = Uint8Array.from(atob(pGpub), char => char.charCodeAt(0));
+    const pGprivUint8Array = Uint8Array.from(atob(pGpriv), char => char.charCodeAt(0));
+    const pGsignUint8Array = Uint8Array.from(atob(pGsign), char => char.charCodeAt(0));
```
WHat are you trying to do here? pG{pub,priv,sign} are base64-encoded, not ascii. Then you should use `Buffer.from(<base64-encoded-string>).toString('base64')` to get the corresponding UInt8Array.

conradoGen replied:
those are initially base64 because were copied from the Playground to make possible to check if a signature validated in the Playground, validates also on nodejs, but while using the same base64, it fails on nodejs while pass on Playground.

the previous portion that has the base64 as output is the opposite, keys and signature are created by nodejs and verify locally, but when the base64 output is copy and pasted to Playground it fails the verification

conradoGen committed:
implemented suggestion and perform signature verification and length comparison

conradoGen commented:
**if the source base64 constants pG{pub, message, sign} are copied and pasted in the Playground the signature pass the verification**

**using any of the conversion below results in a failed verification on nodeJS:**

modified to convert as recommended
`Uint8Array.from(Buffer.from(pGMessage,'base64').toString('base64'), char => char.charCodeAt(0))`
got wrong lengths and verification failed...

added another method to convert 
`new Uint8Array(Buffer.from(pGpub,'base64'))`
despite the correct length, the verification also fail 

the previous conversion approach is kept for comparison 
`Uint8Array.from(atob(pGpub), char => char.charCodeAt(0))`
despite the correct length, the verification also fail

jboillot commented:
Here are the 2 erroneous lines in your modification:
`console.log(Buffer.from(message).toString('base64'));` should be `console.log(Buffer.from(message).toString('ascii'));`
Indeed, the playground is waiting for a plaintext message in ascii, not in base64.
And, for the same reason, `const pGmessageUint8Array = Uint8Array.from(atob(pGMessage), char => char.charCodeAt(0));` should be `const pGmessageUint8Array = Uint8Array.from(Array.from(pGMessage).map(letter => letter.charCodeAt(0)));`.
However, currently you're signing the message `RGFzaGxhbmU=` in ascii. I guess you would instead sign its ascii representation: 'Dashlane'. You just have to replace its value and regenerate a signature, e.g., on the playground.

jboillot commented:
Here is the updated script that does match with the playground:
https://github.com/jboillot/test-dilithium/blob/demonstrate-validity-with-playground/index.js

conradoGen committed:
adjusted examples to validate with playground

conradoGen commented:
thanks for your time... indeed that did the trick! appreciate your work on this!
